src/deploy/functions/iambinder.ts

```diff
@@ -0,0 +1,298 @@
+          [this.defaultComputeServiceAccount(endpoint)]
+        );
+        if (v2Events.isStorageTriggered(endpoint)) {
+          // Storage account associated w/ Cloud Storage must have permission to publish pubsub messages.
+          this.addBinding(
+            `//cloudresourcemanager.googleapis.com/projects/${this.projectNumber}`,
+            "roles/pubsub.publisher",
+            [await this.fetchStorageServiceAccount()]
+          );
+        }
+      }
+    }
+  }
+
+  private addBinding(resource: Resource, role: Role, members: string[]): void {
+    const bindings = this.additions[resource] || new IamBindings();
+    bindings.add(role, members);
+    this.additions[resource] = bindings;
+  }
+
+  private async fetchStorageServiceAccount(): Promise<string> {
+    if (!this.storageServiceAccount) {
+      const resp = await storage.getServiceAccount(this.projectId);
+      this.storageServiceAccount = `serviceAccount:${resp.email_address}`;
+    }
+    return this.storageServiceAccount;
+  }
+
+  private defaultComputeServiceAccount(endpoint: backend.Endpoint): string {
+    if (endpoint.platform === "gcfv1") {
+      return `serviceAccount:${this.projectId}@appspot.gserviceaccount.com`;
+    } else if (endpoint.platform === "gcfv2") {
+      return `serviceAccount:${this.projectNumber}-compute@developer.gserviceaccount.com`;
+    }
+    assertExhaustive(endpoint.platform);
+  }
+
+  static async updatePolicy(resource: Resource, bindings: IamBindings): Promise<iam.Policy> {
+    const existingPolicy = await IamBinder.getPolicy(resource);
+    const existingBindings = IamBindings.fromIamBindings(existingPolicy.bindings);
+
+    const diffBindings = bindings.diff(existingBindings);
+    if (Object.keys(diffBindings.additions).length === 0) {
+      return existingPolicy;
+    }
+
+    const mergedPolicy = {
+      ...existingPolicy,
+      bindings: bindings.merge(existingBindings).asIamBindings(),
+    };
+    await IamBinder.setPolicy(resource, mergedPolicy);
+    return mergedPolicy;
+  }
+
+  static async getPolicy(resource: Resource): Promise<iam.Policy> {
+    // Some GCP APIs use GET for getIamPolicy. Some uses POST. What in the world?
+    if (isSecreteManagerResource(resource)) {
+      const resp = await IamBinder.clientReq<unknown, iam.Policy>("GET", resource, ":getIamPolicy");
+      return resp.body;
+    } else if (isResourceManagerResource(resource)) {
+      const resp = await IamBinder.clientReq<unknown, iam.Policy>(
+        "POST",
+        resource,
+        ":getIamPolicy"
+      );
+      return resp.body;
+    } else {
+      assertExhaustive(resource);
+    }
+  }
+
+  static async setPolicy(resource: Resource, policy: iam.Policy): Promise<iam.Policy> {
+    const resp = await IamBinder.clientReq<{ policy: iam.Policy }, iam.Policy>(
+      "POST",
+      resource,
+      ":setIamPolicy",
+      { policy }
+    );
+    return resp.body;
+  }
+
+  private static clientReq<ReqT, ResT>(
```

Comment:
Since we usually have all of our GCP api calls contained in the `gcp` folder, this kinda seems wrong to me. But I get that we need some special logic to figure out the correct API endpoint, maybe we could keep that here but call out to the resource manager & secret manager APIs?

Reply:
Hmm I see your point.

I was trying to abuse the pattern on that there is always going to be `setIamPolicy` and `getIamPolicy` endpoints for each GCP resources for which the URL can be fully constructed by the full resource name.

The fact that I need to sprinkle on API version and that HTTP method differs for `getIamPolicy` for secret manager and resource manager API ends up making this more complicated than I hoped :(. Going to see what I can do to cleanup here.